feat: Opentelemetry support for metrics and traces

4molybdenum2 opened the pull request:
Signed-off-by: Tathagata Paul <tathagatapaul7@gmail.com>

## Related issue
Closes #3120 
<!--
Please link the GitHub issue this pull request resolves in the format of `#1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## What type of PR is this
/kind enhancement
<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes

#### Metrics
- [x]  Add support for OpenTelemetry in Kyverno to export metrics
- [x] Remove the current Prometheus config
- [x] Configure server-side TLS for exporting metrics using Opentelemetry

#### Traces
- [x] Add initial trace config
- [x] Add tracing for cosign function calls
- [x] Add tracing for webhook function calls
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

WIP PR to track the integration of OpenTelemetry support in Kyverno
<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+		// here the certificates are stored as configmaps
+		secret, err := kubeClient.CoreV1().Secrets("kyverno").Get(ctx, certs, v1.GetOptions{})
+		if err != nil {
+			log.Error(err, "Error fetching certificate from secret")
```
Do we want to continue on errors?

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+		resource.WithSchemaURL(semconv.SchemaURL),
 	)
+	if err != nil {
+		log.Error(err, "failed creating resource")
```
Continue on error?

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
-	admissionReviewDurationLabels := []string{
-		"resource_kind", "resource_namespace", "resource_request_operation",
+	m, err = initializeMetrics(m)
+
```
Suggest removing extra line.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+	config := prometheus.Config{}
+	res, err := resource.New(context.Background(),
+		resource.WithAttributes(semconv.ServiceNameKey.String("kyverno-svc-metrics")),
+		resource.WithAttributes(semconv.ServiceNamespaceKey.String("kyverno")),
```
Needs to load namespace dynamically via `config.KyvernoNamespace`.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+
+	config := prometheus.Config{}
+	res, err := resource.New(context.Background(),
+		resource.WithAttributes(semconv.ServiceNameKey.String("kyverno-svc-metrics")),
```
Is the service name fixed? If so, needs to call ou in the doc.

4molybdenum2 replied:
Service name is fixed for now. Will update in docs

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+	// Create new config object and attach metricsConfig to it
+	m := new(MetricsConfig)
+	m.Config = metricsConfigData
+
```
Remove extra line.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+	m.Log = log
+
+	m, err = initializeMetrics(m)
+
```
Remove extra line.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+
+func (m *MetricsConfig) RecordPolicyResults(policyValidationMode PolicyValidationMode, policyType PolicyType, policyBackgroundMode PolicyBackgroundMode, policyNamespace string, policyName string,
+	resourceKind string, resourceNamespace string, resourceRequestOperation ResourceRequestOperation, ruleName string, ruleResult RuleResult, ruleType RuleType,
+	ruleExecutionCause RuleExecutionCause, log logr.Logger) {
```
`MetricsConfig` has a logger, do we need to pass it in?

4molybdenum2 replied:
Removed this.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+	m.policyResultsMetric.Add(ctx, 1, commonLabels...)
+}
+
+func (m *MetricsConfig) RecordPolicyChanges(policyValidationMode PolicyValidationMode, policyType PolicyType, policyBackgroundMode PolicyBackgroundMode, policyNamespace string, policyName string, policyChangeType string, log logr.Logger) {
```
Remove logger.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+}
+
+func (m *MetricsConfig) RecordPolicyRuleInfo(policyValidationMode PolicyValidationMode, policyType PolicyType, policyBackgroundMode PolicyBackgroundMode, policyNamespace string, policyName string,
+	ruleName string, ruleType RuleType, status string, metricValue float64, log logr.Logger) {
```
Remove logger.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+	m.policyRuleInfoMetric.Observe(ctx, metricValue, commonLabels...)
+}
+
+func (m MetricsConfig) RecordAdmissionRequests(resourceKind string, resourceNamespace string, resourceRequestOperation ResourceRequestOperation, log logr.Logger) {
```
Remove logger.

4molybdenum2 replied:
This will be removed. Marking as resolved.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+
+func (m *MetricsConfig) RecordPolicyExecutionDuration(policyValidationMode PolicyValidationMode, policyType PolicyType, policyBackgroundMode PolicyBackgroundMode, policyNamespace string, policyName string,
+	resourceKind string, resourceNamespace string, resourceRequestOperation ResourceRequestOperation, ruleName string, ruleResult RuleResult, ruleType RuleType,
+	ruleExecutionCause RuleExecutionCause, generalRuleLatencyType string, ruleExecutionLatency float64, log logr.Logger) {
```
Remove logger.

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+	m.policyExecutionDurationMetric.Record(ctx, ruleExecutionLatency, commonLabels...)
+}
+
+func (m *MetricsConfig) RecordAdmissionReviewDuration(resourceKind string, resourceNamespace string, resourceRequestOperation string, admissionRequestLatency float64, log logr.Logger) {
```
Remove logger.

realshuting commented on pkg/tracing/tracing.go:
```diff
@@ -0,0 +1,95 @@
+
+	var client otlptrace.Client
+
+	if certs != "" {
+		// here the certificates are stored as configmaps
+		secret, err := kubeClient.CoreV1().Secrets("kyverno").Get(ctx, certs, v1.GetOptions{})
+		if err != nil {
+			log.Error(err, "Error fetching certificate from secret")
+		}
+
+		cp := x509.NewCertPool()
+		if !cp.AppendCertsFromPEM([]byte(secret.Data["ca.pem"])) {
+			return fmt.Errorf("credentials: failed to append certificates")
+		}
+
+		transportCreds := credentials.NewClientTLSFromCert(cp, "")
+
+		client = otlptracegrpc.NewClient(
+			otlptracegrpc.WithEndpoint(endpoint),
+			otlptracegrpc.WithTLSCredentials(transportCreds),
+		)
+	} else {
+		client = otlptracegrpc.NewClient(
+			otlptracegrpc.WithEndpoint(endpoint),
+			otlptracegrpc.WithInsecure(),
+		)
+	}
```
Duplicate code, can we extract it to a separate function and re-use it?

4molybdenum2 replied:
I was thinking a function ```FetchCert``` in a utils package?

4molybdenum2 replied:
Needed for both tracing and metrics

realshuting reviewed (changes requested):
@4molybdenum2 - sorry for the delay, please see comments.

Also needs doc updates.

4molybdenum2 commented on cmd/kyverno/main.go:
```diff
@@ -79,6 +83,10 @@ func main() {
 	flag.BoolVar(&profile, "profile", false, "Set this flag to 'true', to enable profiling.")
 	flag.StringVar(&profilePort, "profilePort", "6060", "Enable profiling at given port, defaults to 6060.")
 	flag.BoolVar(&disableMetricsExport, "disableMetrics", false, "Set this flag to 'true', to enable exposing the metrics.")
+	flag.BoolVar(&disableTracing, "disableTracing", true, "Set this flag to 'false', to enable exposing traces.")
```
This is a little bit confusing (even for the metrics flag)
By disableTracing = true -> means tracing is disabled. So technically to enable it we have to set it to false

realshuting replied:
Or can we rename it to `enableTracing` and default it to false? This would be clear.

4molybdenum2 commented on cmd/kyverno/main.go:
```diff
@@ -79,6 +83,10 @@ func main() {
 	flag.BoolVar(&disableMetricsExport, "disableMetrics", false, "Set this flag to 'true', to enable exposing the metrics.")
+	flag.BoolVar(&disableTracing, "disableTracing", true, "Set this flag to 'false', to enable exposing traces.")
+	flag.StringVar(&otel, "otelConfig", "grpc", "Set this flag to 'prometheus', to enable exposing metrics directly to prometheus. Or else set to grpc to export metrics to an Opentelemetry collector")
+	flag.StringVar(&otelCollector, "otelCollector", "opentelemetrycollector.kyverno.svc.cluster.local:4317", "Set this flag to the OpenTelemetry Collector Receiver endpoint")
```
How should we do this? Do you mean to separately take service and port values from the user?

realshuting replied:
I was just asking - will there be a conflict if the same port is used by another process?

If so, can we do the same as `profilePort`/`metricsPort`?

4molybdenum2 replied:
Yes that can be possible...let me check and get back to you

4molybdenum2 replied:
I updated it to opentelemetrycollector.kyverno.svc.cluster.local and the metricsPort will be used for the required port on the Opentelemetry Collector (which can be defined by the user).
If we want the user to define their own collector name then we will want to have another input parameter... do we want that?

4molybdenum2 commented on charts/kyverno/values.yaml:
```diff
@@ -142,6 +142,7 @@ envVars: {}
 # -- Extra arguments to give to the binary.
 extraArgs:
   - --autogenInternals=false
+  - --otelConfig=prometheus
```
I will change the default to "prometheus", this will be removed.

4molybdenum2 commented on charts/kyverno/README.md:
```diff
@@ -96,7 +96,7 @@ The command removes all the Kubernetes components associated with the chart and
 | envVarsInit | object | `{}` | Env variables for initContainers. |
 | envVars | object | `{}` | Env variables for containers. |
-| extraArgs | list | `["--autogenInternals=false"]` | Extra arguments to give to the binary. |
+| extraArgs | list | `["--autogenInternals=false","--otelConfig=prometheus"]` | Extra arguments to give to the binary. |
```
This will be removed

chipzoller commented:
@realshuting we need to make sure we're following the KDP process we have outlined. There is still an un-merged PR for this feature [here](https://github.com/kyverno/KDP/pull/16).

4molybdenum2 commented:
Yes, I will need to update the KDP, and also the website documentation for the same. I will do that ASAP.

4molybdenum2 commented on config/manifest/deployment.yaml:
```diff
@@ -78,6 +78,7 @@ spec:
           # - --genWorkers=20
           - "-v=2"
           - --autogenInternals=false
+          - --otelConfig=prometheus
```
No I will remove this

4molybdenum2 replied:
Removed this.

realshuting commented:
> @realshuting Can you check this: https://hackmd.io/@4molybdenum2/kyverno-opentelemetry . This will be added to the documentation portion for this PR

@4molybdenum2 - can you convert it to a PR so that it's easier to track and comment?

4molybdenum2 commented:
Sure

4molybdenum2 commented:
@realshuting made the rest of the suggested changes and rebased branch

realshuting commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,317 @@
+	"time"
 
+	"github.com/go-logr/logr"
 	"github.com/kyverno/kyverno/pkg/config"
```
@4molybdenum2 - missing this.

4molybdenum2 replied:
Ohh, didn't notice....will fix this..

realshuting commented on cmd/kyverno/main.go:
```diff
@@ -79,6 +83,10 @@ func main() {
+	flag.BoolVar(&disableTracing, "disableTracing", true, "Set this flag to 'false', to enable exposing traces.")
+	flag.StringVar(&otel, "otelConfig", "grpc", "Set this flag to 'prometheus', to enable exposing metrics directly to prometheus. Or else set to grpc to export metrics to an Opentelemetry collector")
+	flag.StringVar(&otelCollector, "otelCollector", "opentelemetrycollector.kyverno.svc.cluster.local:4317", "Set this flag to the OpenTelemetry Collector Receiver endpoint")
+	flag.StringVar(&transportCreds, "transportCreds", "", "Set this flag to the CA secret containing the certificate to be be used by our Opentelemetry Metrics Client. If empty string is set, means an insecure connection will be used")
```
Was this updated?

4molybdenum2 replied:
Fixed this

realshuting commented on cmd/kyverno/main.go:
```diff
@@ -235,22 +243,47 @@ func main() {
+			setupLog.Error(err, "Failed to enable tracing for Kyverno")
 			os.Exit(1)
 		}
-		metricsServerMux = http.NewServeMux()
-		metricsServerMux.Handle("/metrics", promhttp.HandlerFor(promConfig.MetricsRegistry, promhttp.HandlerOpts{Timeout: 10 * time.Second}))
```
? ?

4molybdenum2 replied:
No we don't need this because this is being already created in the metrics.go file

4molybdenum2 commented:
@realshuting 

> @4molybdenum2 - to make re-review easier, please resolve all addressed comments.

@realshuting for the first one I updated it to **opentelemetrycollector.kyverno.svc.cluster.local** and the metricsPort will be used for the required port on the Opentelemetry Collector (which can be defined by the user). For the second comment, the metrics server mux is c fetched from the **NewPrometheusconfig** function in metrics/metrics.go

realshuting commented:
Which one is the first comment?

I'm still seeing 20 comments remain open while I believe most of them were already addressed. Can you resolve them?

realshuting commented:
Seeing these ones are open:

https://github.com/kyverno/kyverno/pull/3910#discussion_r885849815
https://github.com/kyverno/kyverno/pull/3910#discussion_r885851815
https://github.com/kyverno/kyverno/pull/3910#discussion_r885876358
https://github.com/kyverno/kyverno/pull/3910#discussion_r885877111
https://github.com/kyverno/kyverno/pull/3910#discussion_r885881401
https://github.com/kyverno/kyverno/pull/3910#discussion_r885881948
https://github.com/kyverno/kyverno/pull/3910#discussion_r885886524
https://github.com/kyverno/kyverno/pull/3910#discussion_r885891912
https://github.com/kyverno/kyverno/pull/3910#discussion_r885893302
https://github.com/kyverno/kyverno/pull/3910#discussion_r885893569
https://github.com/kyverno/kyverno/pull/3910#discussion_r885894292
https://github.com/kyverno/kyverno/pull/3910#discussion_r885894409
https://github.com/kyverno/kyverno/pull/3910#discussion_r885896168
https://github.com/kyverno/kyverno/pull/3910#discussion_r885896437
https://github.com/kyverno/kyverno/pull/3910#discussion_r885896541

4molybdenum2 commented on pkg/cosign/cosign.go:
```diff
@@ -145,9 +146,24 @@ func VerifySignature(opts Options) (digest string, err error) {
+		logger.Info("image verification failed", "error", msg)
+		if strings.Contains(msg, "failed to verify signature") {
+			return "", fmt.Errorf("signature mismatch")
+		} else if strings.Contains(msg, "no matching signatures") {
```
This was fixed by another PR

4molybdenum2 commented on pkg/cosign/cosign.go:
```diff
@@ -145,9 +146,24 @@ func VerifySignature(opts Options) (digest string, err error) {
-		logger.Info("image verification failed", "error", err.Error())
+		msg := err.Error()
+		logger.Info("image verification failed", "error", msg)
+		if strings.Contains(msg, "failed to verify signature") {
```
This was fixed by another PR

4molybdenum2 commented on pkg/metrics/metrics.go:
```diff
@@ -1,143 +1,328 @@
+
+	if certs != "" {
+		// here the certificates are stored as configmaps
+		secret, err := kubeClient.CoreV1().Secrets("kyverno").Get(ctx, certs, v1.GetOptions{})
```
This is fixed

4molybdenum2 commented:
@realshuting I left one unresolved comment for your verification

realshuting commented on pkg/utils/kube/cert.go:
```diff
@@ -0,0 +1,30 @@
+	ctx context.Context,
+	certs string,
+	kubeClient kubernetes.Interface) (credentials.TransportCredentials, error) {
+	secret, err := kubeClient.CoreV1().Secrets(config.KyvernoNamespace()).Get(ctx, certs, v1.GetOptions{})
```
What's the frequency of this client call?

4molybdenum2 replied:
Just one time when kyverno is first started (or restarted...)

realshuting commented on cmd/kyverno/main.go:
```diff
@@ -256,22 +264,49 @@ func main() {
+			metricsAddr := ":" + metricsPort
+			setupLog.Info("Enabling Metrics for Kyverno", "address", metricsAddr)
+
+			endpoint := otelCollector + metricsAddr
+			metricsConfig, err = metrics.NewOTLPGRPCConfig(
+				endpoint,
+				metricsConfigData,
+				transportCreds,
+				kubeClient,
+				log.Log.WithName("Metrics"),
+			)
+
+			if err != nil {
+				setupLog.Error(err, "Failed to enable metrics", "address", ":4317")
+				os.Exit(1)
+			}
+		} else if otel == "prometheus" {
+			// Prometheus Server will serve metrics on metrics-port
+			metricsAddr := ":" + metricsPort
+			setupLog.Info("enabling prometheus metrics", "address", metricsAddr)
+			metricsConfig, metricsServerMux, err = metrics.NewPrometheusConfig(metricsConfigData, log.Log.WithName("Metrics"))
+			go func() {
+				setupLog.Info("Enabling Metrics for Kyverno", "address", metricsAddr)
+				if err := http.ListenAndServe(metricsAddr, metricsServerMux); err != nil {
+					setupLog.Error(err, "failed to enable metrics", "address", metricsAddr)
+					os.Exit(1)
+				}
+			}()
+		}
+	}
+
+	// Tracing Configuration
+	if enableTracing {
+		setupLog.Info("Enabling tracing for Kyverno...")
+		err = tracing.NewTraceConfig(otelCollector, transportCreds, kubeClient, log.Log.WithName("Tracing"))
 		if err != nil {
-			setupLog.Error(err, "failed to setup Prometheus metric configuration")
+			setupLog.Error(err, "Failed to enable tracing for Kyverno")
 			os.Exit(1)
 		}
```
Can we extract the whole metrics logic to a separate file? It can be in `cmd/kyverno/metrics.go` or `pkg/metrics/init.go`.

4molybdenum2 replied:
Yes we can do that...and then call it into main. We can also wait for when the controllers get refactored...WDYT?

realshuting replied:
What controller refactoring?

4molybdenum2 replied:
Sorry I misunderstood this...but yea we can do this.

realshuting reviewed (commented):
Looks good overall!

Do we have scaling testing data?

realshuting reviewed (dismissed):
@4molybdenum2 - sorry for the late response, the PR looks good to me now!

Can you please resolve the conflicts?

@JimBugwadia - any final comments on this?

4molybdenum2 commented:
Faced few conflicts...while fixing I ran ```go mod tidy``` which gave me some errors I couldn't resolve...
But the build seems to be working...WDYT
@realshuting @eddycharly

realshuting reviewed (approved):
/lgtm

Thank you @4molybdenum2 !

4molybdenum2 commented on pkg/tracing/tracing.go:
```diff
@@ -0,0 +1,85 @@
+		return err
+	}
+
+	bsp := sdktrace.NewBatchSpanProcessor(traceExp)
+	// create controller and bind the exporter with it
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithSpanProcessor(bsp),
+		sdktrace.WithResource(res),
+	)
+
+	// set global propagator to tracecontext (the default is no-op).
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetTracerProvider(tp)
+	return nil
```
Ohh right....Yeah, I will create an issue and make a fix tomorrow.
